sdk/src/Services/SSOOIDC/Custom/Internal/_bcl45+netstandard/CoreAmazonSSOOIDC.cs

```diff
@@ -0,0 +1,286 @@
+                }
+                catch (AuthorizationPendingException e)
+                {
+                    // Service is still waiting for user to complete authorization.
+                    // Repeat the loop after an interval.
+                }
+                catch (SlowDownException e)
+                {
+                    // Spec: Add 5 seconds to the polling interval
+                    intervalSec += PollingSlowdownIncrementSeconds;
+                }
+                catch (ExpiredTokenException e)
+                {
+                    // Spec: An exception must be raised, indicating that the SSO login window expired 
+                    // and the SSO login flow must be re-initiated.
+                    throw new AmazonSSOOIDCException("Device code has expired while polling for SSO token, login flow must be re-initiated.", e);
+                }
+                catch (TimeoutException e)
+                {
+                    // Spec: If the call times out then the tool should double its polling interval and then retry.
+                    intervalSec *= 2;
+                }
+                catch (Exception e)
+                {
+                    logger.Error(e, "Unexpected exception while polling for SSO Token.");
+                    Console.WriteLine($"Poll Exception: {e.Message}");
+                    throw;
+                }
+
+                if (DateTime.UtcNow.AddSeconds(intervalSec) > deviceCodeExpiration)
+                {
+                    throw new AmazonSSOOIDCException("Device code has expired while polling for SSO token, login flow must be re-initiated.");
+                }
+
+                context.Sleep(intervalSec * 1000);
+            } // while(polling)
+        }
+#endif
+
+        private static async Task<CreateTokenResponse> PollForSsoTokenAsync(IAmazonSSOOIDC client,
+            CreateTokenRequest createTokenRequest,
+            int pollingIntervalSeconds,
+            DateTime deviceCodeExpiration,
+            IGetSsoTokenContext context)
+        {
+            var logger = Logger.GetLogger(typeof(CoreAmazonSSOOIDC));
+
+            // Spec: If the Interval value is not returned as part of the StartDeviceAuthorization response,
+            // a default Interval value of 5 seconds should be used.
+            var intervalSec = pollingIntervalSeconds > 0 ? pollingIntervalSeconds : DefaultPollingIntervalSeconds;
+
+            // Poll for Token until success, failure, or an error condition arises.
+            while (true)
+            {
+                try
+                {
+                    var response = await client.CreateTokenAsync(createTokenRequest);
+
+                    // If we reach here, the user has completed the SSO Login authorization.
+                    return response;
+                }
+                catch (AuthorizationPendingException e)
+                {
+                    // Service is still waiting for user to complete authorization.
```

Comment:
Technically we are supposed to check if (current time + interval) is later than the previously calculated expiration time.

Reply:
Yup, that happens down on line 277

Reply:
I thought this action was limited to when the error returned was AuthorizationPendingException.  Here, it is being done on every iteration.